### DISCUSSION TITLE
"Assembly with same name is already loaded"

### DISCUSSION THREAD

**Discussion by abbgrade** [PowerShell/PowerShell#21224] [Enhancements]
Hello,

maybe you know the problem already: if you use modules in a script, that rely on the same dotnet assembly, you will sooner or later get the error `Assembly with same name is already loaded`. Not always, since if both instances of the assembly are in the same version, it's fine. If not, then it depends on the order of loading the assemblies. If the newer version is loaded first, then it will work as well. If the older one is loaded first, you cannot import the second module.
The longer you maintain the script, the higher the error probability becomes, since new versions of assemblies will be released and the modules may be updated. So the required order of loading the modules may change.
There are popular assemblies like `Microsoft.Identity.Client`, `Azure.Identity` and some more  that are used by many modules and tend to produce issues.

Have you a workaround or solution for this, that you could share?

In general, I think this is a dotnet / powershell issue and should be fixed. But unfortunately I don't have a working solution approach.

Thanks

**Comment by ThomasNieto** [comment]
You'll want to have the module developers implement an Assembly Load Context for their dependencies so they are not loaded into the default. Here is a PowerShell talk on implementation: https://www.youtube.com/watch?v=KV3D2wP_Uug

**Reply by jborean93** [reply]
I've got a few example repos as well that uses an ALC https://github.com/jborean93/PowerShell-ALC. Unfortunately as a user this doesn't help as it's up to the module author to structure their module to use with an ALC.

**Reply by abbgrade** [reply]
thank you both. I will take some time, that I can have a deeper look, but it looks promising. Maybe I start PRs in my current case: platyPS and powershell-yaml rely on YamlDotNet.

**Reply by jborean93** [reply]
If you are ok with an alternative to `powershell-yaml` you could have a look at my [Yayaml](https://github.com/jborean93/PowerShell-Yayaml) module which loads YamlDotNet in an ALC avoiding any conflicts there. It can be used in conjunction with platyPS as that will continue to load in the normal context while Yayaml uses an ALC to avoid any conflicts with any other module.

**Reply by abbgrade** [reply]
Thanks for the tip. The issue https://github.com/cloudbase/powershell-yaml/issues/121 was already accepted.
